_service/pypi.py: Do not execute pip cache command when a custom directory is provided

JeffreyVdb opened the pull request:
In a scenario where the pip cache directory isn't available because the
user doesn't have permissions; pip will error with this message:

```
ERROR: pip cache commands cannot function since cache is disabled.
```

This currently cannot be mitigated by specifying the `--cache-dir` flag
because the `python -m pip cache dir` command is execute before this
predicate is evaluated. By using an assignment expression (PEP 572), we
can evaluate whether the expression evaluates to a truthy value in the
`elif` branch (saving it's return value) AFTER first evaluating the
custom directory value.

tetsuo-cpp commented on pip_audit/_service/pypi.py:
```diff
@@ -109,12 +109,11 @@ def _get_pip_cache() -> str:
         return str(custom_cache_dir)
-    elif pip_cache_dir is not None:  # pragma: no cover
+    elif (
+        pip_cache_dir := _get_pip_cache() if _PIP_VERSION >= _MINIMUM_PIP_VERSION else None
```
We're trying to support Python 3.6 and later so we'll need to avoid the `:=`.

JeffreyVdb replied:
Fixed in: https://github.com/trailofbits/pip-audit/pull/161/commits/00ead12c134fb0315b4ef6e3026c81823ab0ab7b

tetsuo-cpp reviewed (commented):
Great catch @JeffreyVdb! Thanks for this.

JeffreyVdb commented on pip_audit/_service/pypi.py:
```diff
@@ -109,12 +109,11 @@ def _get_pip_cache() -> str:
         return str(custom_cache_dir)
-    elif pip_cache_dir is not None:  # pragma: no cover
+
+    pip_cache_dir = _get_pip_cache() if _PIP_VERSION >= _MINIMUM_PIP_VERSION else None
```
Typing should be detected here to be `Optional[str]`, but I'm not sure if the convention is to explicitly write it as: 

```python
pip_cache_dir: Optional[str] = _get_pip_cache() if _PIP_VERSION >= _MINIMUM_PIP_VERSION else None
```

Let me know and I'll change it!

woodruffw replied:
Looks good as-is 🙂  -- we don't have a formal style preference for explicit/implicit types, but the general pattern so far has only been to use explicit types when `mypy` needs them.

tetsuo-cpp reviewed (approved):
Nice! This looks good to me.

woodruffw reviewed (approved):
LGTM as well. @JeffreyVdb, would you mind committing a CHANGELOG entry as well, under the "Fixed" heading for the upcoming release? You can use the already present entries as a style reference 🙂

JeffreyVdb commented:
> LGTM as well. @JeffreyVdb, would you mind committing a CHANGELOG entry as well, under the "Fixed" heading for the upcoming release? You can use the already present entries as a style reference 🙂

Rebased and added an entry to the `CHANGELOG`. Hopefully that's sufficient information.

di commented:
Thanks!
